### PR TITLE
cpu/cc2538: add periph/spi driver

### PR DESCRIPTION
**Pull request by hexluthor**
Tested on `cc2538dk`.

### CODE REVIEW

**Comment by haukepetersen** [comment]
Nice. Will look at this in more detail once the feature freeze is lifted and the release is out -> hopefully next week...

One remark: Would you mind remodeling the driver to our more recent style of writing periph drivers, as e.g. in #3856 (for the timer) or in #3838 (for uart). This saves significantly memory...

**Comment by hexluthor** [comment]
Will do.

File: boards/remote/include/periph_conf.h [hunk 1]
```diff
@@ -106,6 +109,32 @@
...
+        .miso_pin = GPIO_PC4,
+        .sck_pin  = GPIO_PD1,
+        .cs_pin   = GPIO_PD3,
+    },
+    {
+        .dev      = SSI1,
+        .mosi_pin = GPIO_PC7,
+        .miso_pin = GPIO_PA4,
+        .sck_pin  = GPIO_PB5,
+    },
```
Comment (hexluthor) [inline boards/remote/include/periph_conf.h]: @alignan which pin serves as MicroSD Chip Select on the ReMote?

**Comment by alignan** [comment]
For the RE-Mote prototype A the CS is grounded so the Micro-SD is always selected, if the Co-Processor enables it first.  Nevertheless we are going to update the RE-Mote source code for the next release just a month away, I will update the Micro-SD CS pin then as we changed the implementation.

**Comment by thomaseichinger** [comment]
@hexluthor I am sorry I didn't respond to this PR recently. I currently can't flash the OpenMote, but I am in contact with the OpenMote people to fix this again. I will test this as soon as I can.

**Comment by alignan** [comment]
I'm starting to work on top of this PR, I will try to post some results over the weekend

**Comment by A-Paul** [comment]
Hi @hexluthor, can you rebase, please?

**Comment by hexluthor** [comment]
@A-Paul rebased!

**Comment by OlegHahm** [comment]
> I'm starting to work on top of this PR, I will try to post some results over the weekend

@alignan, did you find time?

**Comment by kYc0o** [comment]
As soon as the xtimer and i2c PRs are merged, I can test this one too.

**Comment by alignan** [comment]
Not yet, I have a preliminary `CC1200` SPI test for the `RE-Mote` I'm working on, I should have some results soon.

**Comment by A-Paul** [comment]
@hexluthor, I think *cpu/cc2538/include/cc2538_ssi.h* should be named *cpu/cc2538/include/cc2538_spi.h*. Also, there are references and comments with *SSI* or *ssi* at numerous locations.

**Comment by hexluthor** [comment]
@A-Paul, the MCU's internal peripheral is called SSI (Synchronous Serial Interface), and it can be used for SPI as well as a few other protocols. The names in the code match the names in the MCU documentation. `cc2538_ssi.h` describes the general-purpose SSI module and `periph/spi.c` configures it for the more specific SPI mode.

**Comment by A-Paul** [comment]
@hexluthor, definitely a false positive from my side. Thanks for clarifying.
And I thought I found the "mega typo"... :D

**Comment by kYc0o** [comment]
Can we have this ready for this release?

**Comment by hexluthor** [comment]
It's ready now AFAIK.

**Comment by kYc0o** [comment]
I'll test it ASAP

**Comment by kYc0o** [comment]
Can you rebase to add support for mac? #4981

**Comment by hexluthor** [comment]
Rebased onto `master`.

**Comment by A-Paul** [comment]
I had a look with *tests/periph_spi* and a scope on a **cc2538dk** early this day. Only master mode. Looked good so far.
Haven't tested with a logic analyzer or device yet.

**Comment by kYc0o** [comment]
How do you initialise the master device?

**Comment by A-Paul** [comment]
In *periph_spi*?
E.g. `init_master 0 0 13` initializes SPI0 with CS on RF1.17 (PB5). The separate port/pin selection isn't implemented yet, so you have to add 8,16,24... for port A,B,C... to the pin number.

**Comment by A-Paul** [comment]
Did some more testing with *tests/periph_spi* on **cc2538dk** with a logic analyzer.
All 4 modes (combinations of CPOL/CPHA) looking good. Also used 1,5 and 10 MHz clock.

@kYc0o, did you found any issues so far?

**Comment by kYc0o** [comment]
Also tested and works well, I tested all speeds without any issue. So, ACK

**Comment by kYc0o** [comment]
Something is wrong with `spi_transfer_regs`, somehow the macro `PERIPH_SPI_NEEDS_TRANSFER_REGS` is not defined or maybe another problem? it seems this function is defined on `drivers/periph_common/spi.c` but only if that macro is defined.

**Comment by A-Paul** [comment]
@hexluthor, I'm not sure if this can be rated as an easter egg. :-)
Would you please do one of the following:

0. put `#define PERIPH_SPI_NEEDS_TRANSFER_REG` and `#define PERIPH_SPI_NEEDS_TRANSFER_REGS` in *cpu/cc2538/include/periph_cpu.h*
0. implement `spi_transfer_reg` and `spi_transfer_regs` declared by *drivers/include/periph/spi.h*.

The former seems more realistic until freeze on wednesday, I think. :D

**Comment by hexluthor** [comment]
Ok, I think it's fixed. I defined `PERIPH_SPI_NEEDS_TRANSFER_REG`/`REGS` and added `USEMODULE += periph_common`. Thanks.

**Comment by kYc0o** [comment]
Murdock did the job, a big one BTW (24m!) thus, GO!

**Comment by kYc0o** [comment]
Merged #3842

**Pull request merged**